src/main/groovy/com/google/protobuf/gradle/ProtobufPlugin.groovy

```diff
@@ -371,55 +398,49 @@ class ProtobufPlugin implements Plugin<Project> {
-            // 'resources' of the output of 'main', in which the source protos are placed.  This is
-            // nicer than the ad-hoc solution that Android has, because it works for any extended
-            // configuration, not just 'testCompile'.
-            it.inputFiles.from project.sourceSets[sourceSetOrVariantName].compileClasspath
+      return project.tasks.register(extractIncludeProtosTaskName, ProtobufExtract) {
+        it.description = "Extracts proto files from compile dependencies for includes"
+        it.destDir = getExtractedIncludeProtosDir(sourceSetOrVariantName) as File
+        it.inputFiles.from(compileClasspathConfiguration)
+
+        // TL; DR: Make protos in 'test' sourceSet able to import protos from the 'main'
+        // sourceSet.  Sub-configurations, e.g., 'testCompile' that extends 'compile', don't
+        // depend on the their super configurations. As a result, 'testCompile' doesn't depend on
+        // 'compile' and it cannot get the proto files from 'main' sourceSet through the
+        // configuration. However,
+        if (isAndroid) {
+          // TODO(zhangkun83): Android sourceSet doesn't have compileClasspath. If it did, we
+          // haven't figured out a way to put source protos in 'resources'. For now we use an
+          // ad-hoc solution that manually includes the source protos of 'main' and its
+          // dependencies.
+          if (Utils.isTest(sourceSetOrVariantName)) {
+            it.inputFiles.from project.android.sourceSets['main'].proto.sourceDirectories
+            it.inputFiles.from testedCompileClasspathConfiguration
           }
-          it.isTest = Utils.isTest(sourceSetOrVariantName)
+        } else {
+          // In Java projects, the compileClasspath of the 'test' sourceSet includes all the
+          // 'resources' of the output of 'main', in which the source protos are placed.  This is
+          // nicer than the ad-hoc solution that Android has, because it works for any extended
+          // configuration, not just 'testCompile'.
+          it.inputFiles.from project.sourceSets[sourceSetOrVariantName].compileClasspath
         }
+        it.isTest = Utils.isTest(sourceSetOrVariantName)
       }
-
-      generateProtoTask.addIncludeDir(project.files(task))
     }
 
-    private void linkGenerateProtoTasksToTaskName(String compileTaskName, GenerateProtoTask genProtoTask) {
-      Task compileTask = project.tasks.findByName(compileTaskName)
-      if (compileTask != null) {
-        linkGenerateProtoTasksToTask(compileTask, genProtoTask)
-      } else {
+    private void linkGenerateProtoTasksToTaskName(String compileTaskName, Provider<GenerateProtoTask> genProtoTask) {
+      try {
+        project.tasks.named(compileTaskName).configure { compileTask ->
+          linkGenerateProtoTasksToTask(compileTask, genProtoTask)
+        }
+      } catch (UnknownDomainObjectException ignore) {
```

Comment:
https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskCollection.html#named-java.lang.String-
`UnknownDomainObjectException ` -> `UnknownTaskException`

Reply:
UnknownTaskException is an UnknownDomainObjectException. I don't care to use the task-specific APIs when unnecessary, as it means learning 2x the number of APIs while generally not adding value. I'm essentially calling https://docs.gradle.org/current/javadoc/org/gradle/api/NamedDomainObjectCollection.html#named-java.lang.String- here.